File: mavkit-python/examples/list_modes.py
"""List available flight modes for the connected vehicle."""

import asyncio
import os

import mavkit


async def main():
    bind_addr = os.environ.get("MAVKIT_EXAMPLE_UDP_BIND", "0.0.0.0:14550")

    vehicle = await mavkit.Vehicle.connect_udp(bind_addr)
    modes = vehicle.available_modes()
    current_handle = modes.current()
    catalog_handle = modes.catalog()
    identity = vehicle.identity()

    current = current_handle.latest()
    if current is None:
        current = await current_handle.wait()

    print(f"current mode: {current.name} (id={current.custom_mode})\n")

    catalog = catalog_handle.latest()
    if catalog is None:
        catalog = await catalog_handle.wait()

    if not catalog:
        print(f"no mode list available for autopilot {identity.autopilot}")
    else:
        print("available flight modes:")
        for mode in catalog:
            marker = " <-- active" if mode.custom_mode == current.custom_mode else ""
            print(f"  {mode.custom_mode:>3}  {mode.name}{marker}")

    await vehicle.disconnect()


asyncio.run(main())

File: examples/list_modes.rs
use mavkit::Vehicle;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let bind_addr =
        std::env::var("MAVKIT_EXAMPLE_UDP_BIND").unwrap_or_else(|_| "0.0.0.0:14550".to_string());

    let vehicle = Vehicle::connect_udp(&bind_addr).await?;
    let modes = vehicle.available_modes();
    let current_handle = modes.current();
    let catalog_handle = modes.catalog();
    let identity = vehicle.identity();

    let current = if let Some(current) = current_handle.latest() {
        current
    } else {
        current_handle.wait().await?
    };

    println!(
        "current mode: {} (id={})\n",
        current.name, current.custom_mode
    );

    let catalog = if let Some(catalog) = catalog_handle.latest() {
        catalog
    } else {
        catalog_handle.wait().await?
    };

    if catalog.is_empty() {
        println!(
            "no mode list available for autopilot {:?}",
            identity.autopilot
        );
    } else {
        println!("available flight modes:");
        for mode in catalog {
            let marker = if mode.custom_mode == current.custom_mode {
                " <-- active"
            } else {
                ""
            };
            println!("  {:>3}  {}{}", mode.custom_mode, mode.name, marker);
        }
    }

    vehicle.disconnect().await?;
    Ok(())
}
